# TP NoSQL

Prénom :

Nom : 

Pour ce TP noté vous avez à trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [4]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

In [7]:
import pymongo
import pandas as pd

URI = 'mongodb+srv://mongo_user:tNq8TQCQBQb4KrcU@cluster0.pkn17.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0'
client = pymongo.MongoClient(URI)
db = client.tp

# output the name of the collections in the database
print("Collections: ", db.list_collection_names())

Collections:  ['restaurants', 'airbnb']


1. Combien de logements sont la base de données ?

In [9]:
nb = db.airbnb.count_documents({})
print(nb)


5555


In [ ]:
#Il y a 5555 logements dans la base de données.

2. Quel est le prix moyen par ville ? Trier les villes par prix décroissant.

In [102]:
c = db.airbnb.aggregate([
    { "$group": { "_id": "$address.market","prix moyen": { "$avg": "$price" }}},
    { "$sort": { "prix moyen": -1 }}
])

q2 = pd.DataFrame(list(c))
print(q2)


                      _id                           prix moyen
0               Hong Kong  762.4781906300484652665589660743134
1          Rio De Janeiro  525.8059701492537313432835820895522
2   Other (International)                               445.75
3                Istanbul  367.9454545454545454545454545454545
4                   Kauai  288.8208955223880597014925373134328
5                    Maui  286.5882352941176470588235294117647
6                    Oahu  212.2964426877470355731225296442688
7                  Sydney  197.7142857142857142857142857142857
8          The Big Island  179.4604316546762589928057553956835
9                New York  139.6260296540362438220757825370675
10       Other (Domestic)                               128.00
11                                                      115.50
12              Barcelona  100.9477848101265822784810126582278
13               Montreal  100.2330246913580246913580246913580
14                  Porto  69.1281588447653429602888086

In [ ]:
#Voici le prix moyen par ville, Hong Kong est la ville la plus chère.

3. Afficher la liste de tous les différents équipements qui existent.

In [29]:
c = db.airbnb.aggregate([
    { "$unwind": "$amenities" },
    { "$group": { "_id": "$amenities" } },
    { "$project": { "équipments": "$_id", "_id": 0 } }
])

q3 = pd.DataFrame(list(c))
print(q3)

                   équipments
0                     Shampoo
1                        Pool
2                 Double oven
3          Formal dining area
4            Beach essentials
..                        ...
181                   Netflix
182  Pack ’n Play/travel crib
183       Suitable for events
184          24-hour check-in
185                  Elevator

[186 rows x 1 columns]


In [ ]:
#Il existe 185 équipements différents dans les differents logements.

4. Combien de propriétés incluent le Wifi dans les équipements ?

In [37]:
c = db.airbnb.aggregate([
    { "$match": { "amenities": "Wifi" } }, 
    { "$count": "wifi" }  
])

q4= pd.DataFrame(list(c))
print(q4)

   wifi
0  5303


In [ ]:
#il y a 5303 logements qui possedent le wifi

5. Afficher le nom de tous les logements ainsi que le nombre de chambres et de lits qu'ils contiennent (ne pas afficher l'ID)

In [43]:
c = db.airbnb.aggregate([
      { "$group": { "_id": "$name","chambre":{"$sum": "$bedrooms"},"lits": {"$sum": "$beds"}}} 
])

# Conversion en DataFrame pandas
q5 = pd.DataFrame(list(c))

print(q5)

                                                    _id  chambre  lits
0                                       【Ebnb】福田口岸温馨两居室        2     2
1                   Best location in Leblon - 1 bedroom        1     1
2                     Chambre à proximité du métro Parc        1     1
3     5 STARS DOWNTOWN with PARKING ( 2BDR )BELL Center        2     3
4                         MEERBLICK TROPICAL SUITE   CK        0     2
...                                                 ...      ...   ...
5533                 Beautiful Bondi family beach house        4     4
5534                     Economy Single Room in Kumkapi        1     0
5535                            Beach living in Dee Why        2     2
5536                         Jomi House Permit 68030/Al        1     2
5537                 Beautiful Comfy Flat Valencia St A        2     3

[5538 rows x 3 columns]


In [ ]:
#par exemple: "Beautiful Comfy Flat Valencia St A" a 2 chambres et 3 lits


6. Afficher le nom et le prix des logements situés à Porto.

In [59]:
c = db.airbnb.aggregate([
    { "$match": { "address.market": "Porto" } },  
    { "$project": { "nom_du_logement": "$name", "prix": "$price", "_id": 0 } }
])
q6 = pd.DataFrame(list(c))

print(q6)

                         nom_du_logement    prix
0                Ribeira Charming Duplex   80.00
1                      Be Happy in Porto   30.00
2    Downtown Oporto Inn (room cleaning)   40.00
3                         FloresRooms 3T   31.00
4     A Casa Alegre é um apartamento T1.   40.00
..                                   ...     ...
549         Number283 - Your home abroad   48.00
550        Cozy apartment downtown Porto   58.00
551                   Soares Guest House   90.00
552   Central, Modern, University, Metro   41.00
553     StayIN Oporto Música - Apartment  500.00

[554 rows x 2 columns]


In [ ]:
#Il y a 554 logements à Porto, par exemple le logement " Ribeira Charming Duplex" est au prix de $80

7. Quels sont les 5 hôtes les plus populaires (ceux dont les propriétés ont reçu le plus de commentaires) ?

In [75]:
c = db.airbnb.aggregate([
  { "$unwind": "$reviews" },
    { "$group": {"_id": "$name", "commentaires": {"$sum": 1}}},
     { "$sort": { "commentaires": -1 }},
     { "$limit": 5 }
])
q7 = pd.DataFrame(list(c))

print(q7)

                                                 _id  commentaires
0                    #Private Studio - Waikiki Dream           533
1  Near Airport private room, 2 bedroom granny fl...           469
2           La Sagrada Familia (and metro) 4 blocks!           463
3                PRIVATE Room in Spacious, Quiet Apt           420
4                      traditional and Charming room           408


#Celui qui a le plus de commentaires est #Private Studio - Waikiki Dream avec 533 commentaires

8. Quelles sont les 6 villes ayant le plus de logements disponibles à la location ?

In [130]:
c = db.airbnb.aggregate([ 
    { "$group": {"_id": "$address.market", "nb_logement": {"$sum": 1}}},
     { "$sort": {"nb_logement": -1}},
     { "$limit": 6 }
])

q8 = pd.DataFrame(list(c))
print(q8)

         _id  nb_logement
0   Istanbul          660
1   Montreal          648
2  Barcelona          632
3  Hong Kong          619
4     Sydney          609
5   New York          607


In [ ]:
#la ville ayant le plus de logements disponibles à la location est Istanbul.

9. Combien de propriétés acceptent plus de 4 invités et ont une caution de moins de 300€ ?

In [132]:
c = db.airbnb.aggregate([ 
     {"$match": {"accommodates": {"$gt": 4}, "security_deposit": {"$lt": 300}}},
     {"$project": {"name": 1,"security_deposit": 1,"accommodates": 1,"room_type": 1}},
])
q9 = pd.DataFrame(list(c))
print(q9)

          _id                                               name  \
0    10006546                            Ribeira Charming Duplex   
1    10051164                 Catete's Colonial Big Hause Room B   
2     1022200                Kailua-Kona, Kona Coast II 2b condo   
3    10133350                          2 bedroom Upper east side   
4    10166883   Large railroad style 3 bedroom apt in Manhattan!   
..        ...                                                ...   
446   9567840                  Large Comfortable CBD Convenience   
447   9990304     March Madness Special! Aina Nalu Platinum D107   
448   9372696               Economy Family Room for 4-5pax (306)   
449   9612902  1921 15kms from Sydney CBD with Pool Spa parking.   
450   9917859                   StayIN Oporto Música - Apartment   

           room_type  accommodates security_deposit  
0    Entire home/apt             8           200.00  
1       Private room             8             0.00  
2    Entire home/apt 

In [ ]:
# 451 propriétés acceptent plus de 4 invités et ont une caution de moins de 300€


10. Donner les 20 utilisateurs qui ont fait le plus de commentaires (afficher seulement l'ID et le nom de l'utilisateur).

In [124]:
c = db.airbnb.aggregate([
    { "$unwind": "$reviews" },
    { "$group": {"_id": "$reviews.reviewer_id", "nom_utilisateur": { "$first": "$reviews.reviewer_name" },"nombre_de_commentaires": { "$sum": 1 }}},
    { "$sort": { "nombre_de_commentaires": -1 } },
    { "$limit": 20 },
    { "$project": { "_id": 1, "nom_utilisateur": 1 } }
])

q10 = pd.DataFrame(list(c))
print(q10)

          _id nom_utilisateur
0    20775242          Filipe
1    67084875            Nick
2     2961855             Uge
3   162027327           Thien
4    20991911            Lisa
5     1705870           David
6    60816198            Todd
7    12679057            Jodi
8    55241576        Courtney
9    69140895            Lisa
10   78093968           David
11   47303133           Lance
12   57325457            Mary
13   24667379           Karen
14   86665925           Chris
15   25715809           Megan
16   73708321         Gonzalo
17  128210181         Branden
18   61469899            Erik
19   34005800             Dan


In [ ]:
#Filipe,Nick,Uge,Thien et Lisa font partis des 20 utilisateurs qui ont fait le plus de commentaires

11. Parmi les logements à Sydney, quel est la note moyenne des visiteurs ?

In [116]:
c = db.airbnb.aggregate([
    { "$match": { "address.market": "Sydney" } },
    { "$group": { "_id": None,"note_moyenne": { "$avg": "$review_scores.review_scores_rating" }}},
    { "$project": { "_id":"Sydney", "note_moyenne": 1 } }
])

q11 = pd.DataFrame(list(c))
print(q11)



   note_moyenne     _id
0      93.45567  Sydney


In [118]:
#la note moyenne des visiteurs de Sydney est 93.45

12. (bonus) Afficher les logements qui contiennent le mot "park" dans leur nom

13. (bonus) Afficher le nom des logements ayant une lattitude comprise entre 36,1 et 40,6 dont le prix est entre 100 et 200 euros.

In [92]:
 c = db.airbnb.aggregate([
     { "$match": { 
        "address.location.coordinates.1": { "$gte": 36.1, "$lte": 40.6 },
        "price": { "$gte": 100, "$lte": 200 }
    }},
    { "$project": { "nom_du_logement": "$name", "_id": 0,"prix": "$price" }}
    
])
q13 = pd.DataFrame(list(c))

print(q13)

                                  nom_du_logement    prix
0              Cozy Comfy Suite of Rockaway Beach  125.00
1       Coney Island MCU Park Wi fi Cable Apt****  110.00
2  Sheepshead Bay, 1-bdr apartment. Close to all!  130.00


In [ ]:
#il y a 3 logements  ayant une lattitude comprise entre 36,1 et 40,6 dont le prix est entre 100 et 200 euros: Cozy Comfy Suite of Rockaway Beach,Coney Island MCU Park Wi fi Cable Apt**** et Sheepshead Bay, 1-bdr apartment. Close to all!